wp/lib/bap_wp/src/output.ml

```diff
@@ -0,0 +1,126 @@
+
+let string_of_sign s =
+  match s with
+  | Unsigned -> "u"
+  | Signed -> "s"
+
+exception NoValue of string
+
+let no_value_msg e m =
+  let e_str = Z3.Expr.to_string e in
+  let m_str = Z3.Model.to_string m in
+  Printf.sprintf "No value for symbol '%s' found in model: %s" e_str m_str
+
+let word_of_str s w sign =
+  let w_str = Printf.sprintf "%s:%d%s" s w (string_of_sign sign) in
+  Word.of_string w_str
+
+let z3_val_of e m =
+  match Z3.Model.eval m e true with
+  | Some v -> v
+  | None -> raise (NoValue (no_value_msg e m))
+
+let word_of_z3_bv bv =
+  let bits = Z3.BitVector.get_size (Z3.Expr.get_sort bv) in
+  let sign = Unsigned in
+  let s = Z3.Expr.to_string bv in
+  word_of_str (String.substr_replace_first s ~pattern:"#" ~with_:"0") bits sign
+
+let val_for_reg r m env =
+  let reg_var = Var.create r (Type.imm 64) in
+  let reg_value, _ = Environment.get_var env reg_var in
+  let v = z3_val_of reg_value m in
+  word_of_z3_bv v
+
+let strip_final_digits s =
+  let s' = List.rev (String.to_list s) in
+  let rec pop_digits l =
+    match l with
+    | [] -> l
+    | x :: xs ->
+      begin
+        match x with
+        | '0' .. '9' -> pop_digits xs
+        | _ -> l
+      end
+    in 
+  let s'' = List.rev (pop_digits s') in
+  String.of_char_list s''
+
+let reg_names m =
+  m |>
+  Z3.Model.get_decls |>
+  List.map ~f:Z3.FuncDecl.get_name |>
+  List.map ~f:Z3.Symbol.to_string |>
+  List.filter ~f:(fun s -> String.is_substring_at s 0 "R") |>
+  List.map ~f:strip_final_digits
+
+
+let print_result (solver : Z3.Solver.solver) (status : Z3.Solver.status)
+    (goals: Constr.t) (ctx : Z3.context) : unit =
+  match status with
+  | Z3.Solver.UNSATISFIABLE -> Format.printf "\nUNSAT!\n%!"
+  | Z3.Solver.UNKNOWN -> Format.printf "\nUNKNOWN!\n%!"
+  | Z3.Solver.SATISFIABLE ->
+    Format.printf "\nSAT!\n%!";
+    let model = Z3.Solver.get_model solver
+                |> Option.value_exn ?here:None ?error:None ?message:None in
+    Format.printf "\nModel:\n%s\n%!" (Z3.Model.to_string model);
+    let refuted_goals = Constr.get_refuted_goals goals solver ctx in
+    Format.printf "\nRefuted goals:\n%!";
+    Seq.iter refuted_goals ~f:(fun g ->
+        Format.printf "%s\n%!" (Constr.refuted_goal_to_string g model))
+
+(** [output_gdb] is similar to [print_result] except chews on the model and outputs a gdb script with a 
+    breakpoint at the subroutine and fills the appropriate registers *)
+
+let output_gdb (solver : Z3.Solver.solver) (status : Z3.Solver.status)
+               (env : Env.t) ~func:(func : string) ~filename:(gdb_filename : string) : unit =
+    match status with
+    | Z3.Solver.SATISFIABLE ->
+        let model = Z3.Solver.get_model solver
+                |> Option.value_exn ?here:None ?error:None ?message:None in
+        let regs = reg_names model in
+        Out_channel.with_file gdb_filename  ~f:(fun t -> 
+            Printf.fprintf t "break *%s\n" func; (* The "*" is necessary to break before some slight setup *)
```

Comment:
Not super important, but I noticed we use both the `Printf` modules and the `Format` modules. What's the difference between these two and would it be a good idea to be uniform?

Reply:
I have not figured out how to get Format to actually print to a file. Me and jt scratched our heads about this one and couldn't figure it out. 

```ocaml
        Out_channel.with_file gdb_filename  ~f:(fun t -> 
            let t = Format.formatter_of_out_channel t in
            Format.fprintf t "break *%s\n" func; (* The "*" is necessary to break before some slight setup *)
            Format.fprintf t "start\n";
            List.iter regs ~f:(fun r -> 
                  let w = val_for_reg r model env in
                  let s = Format.asprintf "set $%s = %a \n" (String.lowercase r) Bap.Std.Word.pp_hex w  in
                  Format.fprintf t "%s" s))   
```

outputs a blank file. Thoughts? Printf was a workaround.

Reply:
Tried some various flush commands too.

Reply:
That's super interesting. Flushing sounds like the intuitive answer to this. If that's the case, keeping it as `Printf` is all good with me.